lib/pure/asyncdispatch.nim

```diff
@@ -758,20 +758,23 @@ when defined(windows) or defined(nimdoc):
       let setoptRet = setsockopt(clientSock, SOL_SOCKET,
           SO_UPDATE_ACCEPT_CONTEXT, addr listenSock,
           sizeof(listenSock).SockLen)
-      if setoptRet != 0: raiseOSError(osLastError())
-
-      var localSockaddr, remoteSockaddr: ptr SockAddr
-      var localLen, remoteLen: int32
-      getAcceptExSockaddrs(addr lpOutputBuf[0], dwReceiveDataLength,
-                           dwLocalAddressLength, dwRemoteAddressLength,
-                           addr localSockaddr, addr localLen,
-                           addr remoteSockaddr, addr remoteLen)
-      register(clientSock.AsyncFD)
-      # TODO: IPv6. Check ``sa_family``. http://stackoverflow.com/a/9212542/492186
-      retFuture.complete(
-        (address: $inet_ntoa(cast[ptr Sockaddr_in](remoteSockAddr).sin_addr),
-         client: clientSock.AsyncFD)
-      )
+      if setoptRet != 0:
+        let errcode = osLastError()
+        discard clientSock.closeSocket()
```

Comment:
What is the semantics of `doAssert`? I think of it as "this must always be true as long as the program is written correctly". This operation can fail even if the program is correct (e.g. on Windows "WSAENETDOWN" can be returned by `close` which means "some network subsystem failure").
So do you still think `doAssert` would be the best option here?

There is also a similar situation in error handling within ioselectors: https://github.com/nim-lang/Nim/blob/devel/lib/pure/ioselects/ioselectors_kqueue.nim#L103
So we should define some pattern for these scenarios and use it consistently.